core/src/main/java/hudson/Launcher.java

```diff
@@ -383,9 +393,24 @@ public Proc start() throws IOException {
 
         /**
          * Starts the process and waits for its completion.
+         * This method stores internal context and hence it is synchronized
+         * @return Return code of the invoked process
+         * @throws IOException Operation error (e.g. remote call failure)
+         * @throws InterruptedException The process has been interrupted
          */
-        public int join() throws IOException, InterruptedException {
-            return start().join();
+        public synchronized int join() throws IOException, InterruptedException {
```

Comment:
I think what we really want to do is:

```
Proc proc = start();
try {
  return proc.join();
} finally {
  if (proc.isAlive()) { // will never happen but this forces Proc to not be removed and early GC by escape analysis
    LOGGER.log(Level.WARNING, "Process not finished after call to join() completed");
  }
}
```

Which has the advantage of reducing the memory footprint and localizing the hack to just a single method